File: lintcode/cpp/Binary_Tree_Preorder_Traversal.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */

class Solution {
public:
    /**
     * @param root: The root of binary tree.
     * @return: Preorder in vector which contains node values.
     */
    vector<int> preorderTraversal(TreeNode *root) {
        // write your code here
        if(!root) return {};
        
        stack<TreeNode *> st;
        st.push(root);
        
        vector<int> result;
        while(!st.empty()){
            TreeNode *t = st.top();
            st.pop();
            
            result.push_back(t->val);
            
            if(t->right) st.push(t->right);
            if(t->left) st.push(t->left);
        }
        
        return result;
    }
};

File: lintcode/java/BinaryTreePreorderTraversal.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: Preorder in ArrayList which contains node values.
     */
    public List<Integer> preorderTraversal(TreeNode root) {
        if(root == null) return Collections.emptyList();
        
        Deque<TreeNode> stack = new ArrayDeque<TreeNode>();
        stack.push(root);
        
        List<Integer> results = new ArrayList<>();
        while(!stack.isEmpty()) {
            TreeNode t = stack.pop();
            results.add(t.val);
            if(t.right != null) stack.push(t.right);
            if(t.left != null) stack.push(t.left);
        }
        
        return Collections.unmodifiableList(results);
    }
}
